src/iceberg/avro/avro_data_util.cc

```diff
@@ -17,16 +17,440 @@
+                           const FieldProjection& projection, const ListType& list_type,
+                           ::arrow::ArrayBuilder* array_builder) {
+  if (avro_node->type() != ::avro::AVRO_ARRAY) {
+    return InvalidArgument("Expected Avro array, got type: {}", ToString(avro_node));
+  }
+  const auto& avro_array = avro_datum.value<::avro::GenericArray>();
+
+  auto* list_builder = internal::checked_cast<::arrow::ListBuilder*>(array_builder);
+  ICEBERG_ARROW_RETURN_NOT_OK(list_builder->Append());
+
+  const auto& element_projection = projection.children[0];
+  auto* value_builder = list_builder->value_builder();
+  const auto& element_node = avro_node->leafAt(0);
+  const auto& element_field = list_type.fields().back();
+
+  for (const auto& element : avro_array.value()) {
+    ICEBERG_RETURN_UNEXPECTED(AppendFieldToBuilder(
+        element_node, element, element_projection, element_field, value_builder));
+  }
+  return {};
+}
+
+/// \brief Append Avro map data to Arrow map builder.
+Status AppendMapToBuilder(const ::avro::NodePtr& avro_node,
+                          const ::avro::GenericDatum& avro_datum,
+                          const FieldProjection& key_projection,
+                          const FieldProjection& value_projection,
+                          const MapType& map_type, ::arrow::ArrayBuilder* array_builder) {
+  auto* map_builder = internal::checked_cast<::arrow::MapBuilder*>(array_builder);
+
+  if (avro_node->type() == ::avro::AVRO_MAP) {
+    // Handle regular Avro map: map<string, value>
+    const auto& avro_map = avro_datum.value<::avro::GenericMap>();
+    const auto& map_entries = avro_map.value();
+
+    const auto& key_node = avro_node->leafAt(0);
+    const auto& value_node = avro_node->leafAt(1);
+
+    const auto& key_field = map_type.key();
+    const auto& value_field = map_type.value();
+
+    ICEBERG_ARROW_RETURN_NOT_OK(map_builder->Append());
+    auto* key_builder = map_builder->key_builder();
+    auto* item_builder = map_builder->item_builder();
+
+    for (const auto& entry : map_entries) {
+      ICEBERG_RETURN_UNEXPECTED(AppendFieldToBuilder(
+          key_node, entry.first, key_projection, key_field, key_builder));
+      ICEBERG_RETURN_UNEXPECTED(AppendFieldToBuilder(
+          value_node, entry.second, value_projection, value_field, item_builder));
+    }
+
+    return {};
+  } else if (avro_node->type() == ::avro::AVRO_ARRAY && HasMapLogicalType(avro_node)) {
+    // Handle array-based map: list<struct<key, value>>
+    const auto& avro_array = avro_datum.value<::avro::GenericArray>();
+    const auto& array_entries = avro_array.value();
+
+    const auto& key_field = map_type.key();
+    const auto& value_field = map_type.value();
+
+    ICEBERG_ARROW_RETURN_NOT_OK(map_builder->Append());
+    auto* key_builder = map_builder->key_builder();
+    auto* item_builder = map_builder->item_builder();
+
+    const auto& record_node = avro_node->leafAt(0);
```

Comment:
should check leafCount == 1?

Reply:
`::avro::AVRO_ARRAY` type should guarantee this so I don't think we need to over-protect it.